lib/cmdline.c

```diff
@@ -1165,11 +1165,11 @@ void rm_cmd_set_paths_from_cmdline(
     g_assert(v->paths);
 
     for(char *path, **list = v->paths; (path = *list); ++list) {
-        if(strcmp(path, "-") == 0) {
+        if(path[0] == '-' && path[1] == 0) {
```

Comment:
The following reply is taken from the log message of the updated commit.

-----------

The way `rmlint` currently works is that it produces a shell script that sometimes performs its work by running `rmlint` many times with varying arguments; `rmlint` can be run a large number of times, and each invocation involves argument processing whereby each path argument undergoes multiple char-by-char comparisons in search of a match for a special, well-known value (e.g., "-" or "//").

Therefore, this commit optimizes the process by using tailored, idiomatic C-string calculations in lieu of invoking a generic, open-ended, arbitrary comparison function; this commit replaces the Standard C Library's `strcmp()` with boolean operators and character comparisons.

The calculations are idiomatic in that they are a familiar way of processing strings, but are somewhat particular to C code. To aid reading this code, comments have been embedded to clarify the purpose, as in the following:

    if(/* path is "-" */ path[0] == '-' && path[1] == 0)

It's true that an optimizing compiler might perform this substitution automatically, or it might even make a better, platform-specific optimization. However, this seems unlikely based on the following rudimentary benchmarks; first define some Bash functions:

    # Copy this script with the following command line, where "$THIS"
    # expands to the revision of this commit log message:
    #
    #   git log -1 --format=%b "$THIS" |
    #     sed -n '/^  #-/,/^  #-/p' | cut -c3- >/tmp/script.sh
    #
    #----------------------------------------------------------------
    make_test()
    {
    	# Note that the body of this function is
    	# explicitly indented with tab characters
    	# for the benefit of the "<<-" operator.

    	local name="$1" expression="$2"

    	cat >"$name".c <<-EOF
    	#include <stdbool.h> // bool
    	#include <string.h>  // strcmp

    	#define ITERATIONS 10000000 // 10 million

    	int main()
    	{
    	    int result = 0;
    	    char string[3] = {0};
    	    const char *const volatile pointer = string;
    	    bool toggle = 0;
    	    for (long i = 0; i < ITERATIONS; ++i) {
    	        string[0] = string[1] = '/' + toggle;
    	        toggle = !toggle;
    	        const char *const s = pointer;
    	        result += $expression;
    	    }
    	    return result;
    	}
    	EOF
    }

    build()
    {
      local name="$1"; shift
      gcc -std=c99 -pedantic -Wall -static -fno-plt -fno-pie "$@" \
        "$name".c -o "$name"
    }

    build_all() { build strcmp "$@" && build equals "$@"; }

    run() { time -p ./"$1"; }

    benchmark()
    {
      echo =========== "$@"
      run strcmp
      echo -----------
      run equals
    }

    build_and_benchmark()
    {
      build_all "$@" || return 1
      benchmark "$@"; return 0
    }
    #----------------------------------------------------------------

Then, create and run some benchmarks:

    make_test strcmp 'strcmp(s,"//")'
    make_test equals "(s[0] == '/' && s[1] == '/' && s[2] == 0)"

    gcc --version | head -n 1
    build_and_benchmark -O0
    build_and_benchmark -Og
    build_and_benchmark -O1
    build_and_benchmark -O2

All together, this produces the following output on my [very old and slow] system; at each optimization level, timing information is provided first for `strcmp()` and then for the manual calculation.  Perversely, optimization seems to make `strcmp()` slower, not faster; in contrast, optimization drastically improves the performance of manual calculation, which is also always faster than using `strcmp()`.

    gcc (GCC) 7.3.1 20180312
    =========== -O0
    real 0.35
    user 0.35
    sys 0.00
    -----------
    real 0.24
    user 0.24
    sys 0.00
    =========== -Og
    real 0.58
    user 0.57
    sys 0.00
    -----------
    real 0.15
    user 0.15
    sys 0.00
    =========== -O1
    real 0.60
    user 0.59
    sys 0.00
    -----------
    real 0.16
    user 0.16
    sys 0.00
    =========== -O2
    real 0.58
    user 0.57
    sys 0.00
    -----------
    real 0.07
    user 0.07
    sys 0.00

Assembler code for the unoptimized case can be produced with the following:

    build_all -S -O0
    cat strcmp # review
    cat equals

Here is the meat of the comparison:

             strcmp                    equals
    ------------------------- | -------------------------
       .LC0:                  |  movl    -20(%ebp), %eax
       .string "//"           |  movzbl  (%eax), %eax
       ...                    |  cmpb    $47, %al           // s[0] == '/'
       pushl   $.LC0          |  jne     .L3
       pushl   -20(%ebp)      |  movl    -20(%ebp), %eax
       call    *strcmp@GOT    |  addl    $1, %eax
                              |  movzbl  (%eax), %eax
                              |  cmpb    $47, %al           // s[1] == '/'
                              |  jne     .L3
                              |  movl    -20(%ebp), %eax
                              |  addl    $2, %eax
                              |  movzbl  (%eax), %eax
                              |  testb   %al, %al           // s[2] == 0
                              |  jne     .L3
                              |  movl    $1, %eax
                              |  jmp     .L4
                              |  .L3:
                              |  movl    $0, %eax

As you can see, one is calling a function, while the other is performing the manual calculations. Now, consider the optimized version:

    build_all -S -O2
    cat strcmp
    cat equals

This yields:

             strcmp                    equals
    ------------------------- | -------------------------
     .LC0:                    |  movl    -20(%ebp), %ebx
     .string "//"             |  cmpb    $47, (%ebx)
     ...                      |  jne     .L2
     movl    -36(%ebp), %esi  |  cmpb    $47, 1(%ebx)
     movl    $.LC0, %edi      |  jne     .L2
     ...                      |  cmpb    $1, 2(%ebx)
     movl    $3, %ecx         |
     repz cmpsb               |
     seta    %cl              |
     setb    %bl              |
     subl    %ebx, %ecx       |
     movsbl  %cl, %ecx        |

There is no longer a call to `strcmp()`, but the compiler has substituted what is essentially an exact replica of `strcmp()` in terms of x86 assember code; the `repz cmpsb` iterates through 2 strings in memory, comparing the bytes, stopping when there is a mismatch. Strangely, one would think that a native x86 instruction sequence might be inherently faster, but the benchmarks reveal that the naive integer comparisons are much faster. I suspect the reason for this is 2-fold:

* Perhaps, `repz cmpsb` is performing a lot more memory access; in contrast, each naive integer comparison instruction can encode a comparand directly (e.g., the number 47, which is ASCII for '/').

* Maybe, `repz cmpsb` cannot benefit from instruction pipelining as readily as separate integer comparisons.

Perhaps the GNU GCC team should re-consider how `strcmp()` is optimized.

Anyway, the results are clear: It's better to do these small calculations by hand, rather than rely on the compiler to do something magical.